rfcs/0161-intelligent-worker-cycling.md

```diff
@@ -0,0 +1,112 @@
+# RFC 0161 - Intelligent worker cycling on worker pool update
+* Comments: [#161](https://github.com/taskcluster/taskcluster-rfcs/pull/161)
+* Proposed by: @petemoore
+
+# Summary
+
+Workers of a worker pool should be gracefully and automatically refreshed or
+decommissioned if they are no longer consistent with the most recent worker
+pool definition.
+
+## Motivation
+
+Without checks and balances in place, worker pools may contain outdated workers
+whose configuration no longer reflects the most recent worker pool definition,
+for an unbounded period of time.
+
+With intelligent worker cycling, we can ensure that workers with outdated
+configuration are swiftly removed from worker pools.
+
+# Details
+
+The responsibility for keeping workers up-to-date is split between Worker
+Manager (for ensuring that provider launch configuration of workers is valid)
+and the workers, responsible for refreshing `workerConfig` on reregisterWorker
+calls.
+
+This RFC involves changes to both Worker Manager and Worker-Managed-spawned
+workers (via Worker Runner):
```

Comment:
It sounds like we're going to apply these proposals to both dynamic cloud and static/hardware instances. Is that accurate?

I think this proposal could be very useful in preventing cloud workers from running with the wrong configuration. Big thumbs up for that use case.

There are a few cases where I have concerns:

- k8s scriptworkers. We're thinking about potentially implementing a canary deployment if a change is backward-compatible, so we can verify the new docker image works in production at a small percentage rollout before completely cutting over. The proposed behavior could conflict with this canary deployment, depending on the contents of `workerConfig`.
- hardware scriptworkers. We would most likely deploy any pool changes via puppet, here. The proposed behavior might result in unnecessary task hiccups, or even worse, idle hardware hosts with no running worker.
- hardware testers. We could potentially deploy pool changes, as described in the proposal. However, a mismatch between puppet and Worker Manager could still render our hardware pools useless.

I think there are various approaches we could take here. We could limit what goes into `workerConfig` on these workers, so the config changes rarely. We may be able to detect and alert on configuration mismatches and idle hardware workers. We could put in a check into ci-admin to prevent errant deploys of mismatching configurations. Any thoughts as to whether the RFC should include one or some of those options?